lib/puppet/pops/puppet_stack.rb

```diff
@@ -1,57 +1,62 @@
-module Puppet::Pops
-# Module for making a call such that there is an identifiable entry on
-# the ruby call stack enabling getting a puppet call stack
-# To use this make a call with:
-# ```
-# Puppet::Pops::PuppetStack.stack(file, line, receiver, message, args)
-# ```
-# To get the stack call:
-# ```
-# Puppet::Pops::PuppetStack.stacktrace
-#
-# When getting a backtrace in Ruby, the puppet stack frames are
-# identified as coming from "in 'stack'" and having a ".pp" file
-# name.
-# To support testing, a given file that is an empty string, or nil
-# as well as a nil line number are supported. Such stack frames
-# will be represented with the text `unknown` and `0´ respectively.
-#
-module PuppetStack
-  # Pattern matching an entry in the ruby stack that is a puppet entry
-  PP_ENTRY_PATTERN = /^(.*\.pp)?:([0-9]+):in (`stack'|`block in call_function'|`<eval>')/
+require 'puppet/thread_local'
 
-  # Sends a message to an obj such that it appears to come from
-  # file, line when calling stacktrace.
-  #
-  def self.stack(file, line, obj, message, args, &block)
-    file = '' if file.nil?
-    line = 0 if line.nil?
+module Puppet
+  module Pops
+    # Utility class for keeping track of the "Puppet stack", ie the file
+    # and line numbers of Puppet Code that created the current context.
+    #
+    # To use this make a call with:
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.stack(file, line, receiver, message, args)
+    # ```
+    #
+    # To get the stack call:
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.stacktrace
+    # ```
+    #
+    # or
+    #
+    # ```rb
+    # Puppet::Pops::PuppetStack.top_of_stack
+    # ```
+    #
+    # To support testing, a given file that is an empty string, or nil
+    # as well as a nil line number are supported. Such stack frames
+    # will be represented with the text `unknown` and `0´ respectively.
+    class PuppetStack
+      @@stack = Puppet::ThreadLocal.new { Array.new }
```

Comment:
I don't like the class variable here, but wasn't sure if we should tie this to the context

Reply:
Why not just a regular class instance variable though

Reply:
If I had a good reason. I don't remember it. What do you think about class vs module (see https://github.com/puppetlabs/puppet/pull/7754#discussion_r330304454)? If the answer is module that might remove the class vs class instance question?

Reply:
Moved this to a "class instance variable" or "module instance variable" or whatever it's called.